webknossos-tracingstore/app/com/scalableminds/webknossos/tracingstore/controllers/VolumeTracingController.scala

```diff
@@ -203,6 +219,159 @@ class VolumeTracingController @Inject()(val tracingService: VolumeTracingService
+              "compressor" -> compressor,
+              "filters" -> None,
+              "shape" -> List(
+                channels,
+                // Zarr can't handle data sets that don't start at 0, so we extend shape to include "true" coords
+                (tracing.boundingBox.width + tracing.boundingBox.topLeft.x) / magParsed.x,
+                (tracing.boundingBox.height + tracing.boundingBox.topLeft.y) / magParsed.y,
+                (tracing.boundingBox.depth + tracing.boundingBox.topLeft.z) / magParsed.z
+              ),
+              "dimension_seperator" -> "."
+            ))
+      }
+  }
+
+  def zGroup(token: Option[String], tracingId: String): Action[AnyContent] = Action.async { implicit request =>
+    accessTokenService.validateAccess(UserAccessRequest.readTracing(tracingId), urlOrHeaderToken(token, request)) {
+      Future(Ok(Json.obj("zarr_format" -> 2)))
+    }
+  }
+
+  def rawZarrCube(token: Option[String], tracingId: String, mag: String, cxyz: String): Action[AnyContent] =
+    Action.async { implicit request =>
+      {
+        val combinedToken = urlOrHeaderToken(token, request)
+        accessTokenService.validateAccess(UserAccessRequest.readTracing(tracingId), combinedToken) {
+          for {
+            tracing <- tracingService.find(tracingId) ?~> Messages("tracing.notFound")
+
+            existingMags = tracing.resolutions.map(v => Vec3Int(v.x, v.y, v.z))
+            magParsed <- Vec3Int.fromMagLiteral(mag, allowScalar = true) ?~> Messages("dataLayer.invalidMag", mag)
+            _ <- bool2Fox(existingMags.contains(magParsed)) ?~> Messages("tracing.wrongMag", tracingId, mag) ~> 404
+
+            (c, x, y, z) <- ZarrCoordinatesParser.parseDotCoordinates(cxyz) ?~> Messages("zarr.invalidChunkCoordinates") ~> 404
+            _ <- bool2Fox(c == 0) ~> Messages("zarr.invalidFirstChunkCoord") ~> 404
+            cubeSize = DataLayer.bucketLength
+            request = WebKnossosDataRequest(
+              position = Vec3Int(x * cubeSize * magParsed.x, y * cubeSize * magParsed.y, z * cubeSize * magParsed.z),
+              mag = magParsed,
+              cubeSize = cubeSize,
+              fourBit = Some(false),
+              applyAgglomerate = None,
+              version = None
+            )
+            (data, _) <- tracingService.data(tracingId, tracing, List(request))
+
+            dataWithFallback <- getFallbackLayerDataIfEmpty(tracing, data, mag, cxyz, combinedToken) ?~> "Getting fallback layer failed" ~> 404
+          } yield Ok(dataWithFallback).withHeaders()
+        }
+      }
+    }
+
+  private def getFallbackLayerDataIfEmpty(tracing: VolumeTracing,
+                                          data: Array[Byte],
+                                          mag: String,
+                                          cxyz: String,
+                                          urlToken: Option[String]): Fox[Array[Byte]] = {
+    val isAnnotationEmpty = data.forall(x => x == 0)
```

Comment:
I think the all-zero test is not the best way to go here. Consider the case that the fallback layer contains some colored data in a bucket, but the user “corrects” that data by erasing it, resulting in an *intended* all-zero bucket. 

Instead, please use the indices array returned by the `tracingService.data` function. It contains the list of buckets (that were requested) that have not been touched in the volume annotation. Since you request only one bucket, you can simply check if this list is empty.